service/collector/update.go

```diff
@@ -0,0 +1,222 @@
+		[]string{
+			"cluster_id",
+			"node_pool_id",
+		},
+		nil,
+	)
+)
+
+type UpdateConfig struct {
+	Helper *helper
+	Logger micrologger.Logger
+}
+
+type Update struct {
+	helper *helper
+	logger micrologger.Logger
+}
+
+type updateInfo struct {
+	nodePoolID      string
+	clusterID       string
+	pauseTime       float64
+	batchPercentage float64
+	batchNumber     float64
+}
+
+func NewUpdate(config UpdateConfig) (*Update, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	np := &Update{
+		helper: config.Helper,
+		logger: config.Logger,
+	}
+
+	return np, nil
+}
+
+func (np *Update) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+
+	var list infrastructurev1alpha2.AWSMachineDeploymentList
+	{
+		err := np.helper.clients.CtrlClient().List(
+			ctx,
+			&list,
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+	var nodePools []updateInfo
+	{
+		for _, md := range list.Items {
+			nodePool := updateInfo{
+				nodePoolID: md.Labels[label.MachineDeployment],
+				clusterID:  md.Labels[label.Cluster],
+			}
+			batch, pauseTime, err := np.getUpdateAnnotations(md)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			//check whether max batch size is given as percentage or integer and calculate the other accordingly
+			if n, err := strconv.Atoi(batch); err == nil {
+				nodePool.batchNumber = float64(n)
+				nodePool.batchPercentage = float64(n / md.Spec.NodePool.Scaling.Min)
+			} else if p, err := strconv.ParseFloat(batch, 64); err == nil {
+				nodePool.batchNumber = float64(md.Spec.NodePool.Scaling.Max) * p
+				nodePool.batchPercentage = p
+			} else if err != nil {
+				return microerror.Mask(err)
+			}
+
+			// calculate the pause time in seconds
+			duration, err := duration.ParseISO8601(pauseTime)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			nodePool.pauseTime = float64(duration.Shift(time.Now()).Second())
+		}
+
+		for _, np := range nodePools {
+			ch <- prometheus.MustNewConstMetric(
+				updateBatchPercentage,
+				prometheus.GaugeValue,
+				np.batchPercentage,
+				np.clusterID,
+				np.nodePoolID,
+			)
+
+			ch <- prometheus.MustNewConstMetric(
+				updateBatchNumber,
+				prometheus.GaugeValue,
+				np.batchNumber,
+				np.clusterID,
+				np.nodePoolID,
+			)
+
+			ch <- prometheus.MustNewConstMetric(
+				updatePauseTime,
+				prometheus.GaugeValue,
+				float64(np.batchNumber),
+				np.clusterID,
+				np.nodePoolID,
+			)
+		}
+	}
+
+	return nil
+}
+
+func (np *Update) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- updateBatchPercentage
+	ch <- updateBatchNumber
+	ch <- updatePauseTime
+
+	return nil
+}
+
+func (np *Update) getUpdateAnnotations(md infrastructurev1alpha2.AWSMachineDeployment) (string, string, error) {
```

Comment:
This function gets the update batch size and update wait time we have set using annotations on either the node pool or the cluster. If the annotations are not set it will return the default values.